### ISSUE TITLE
Use QUIC Session Resumption for 0RTT handshakes

### ISSUE THREAD

**Issue by Stebalien** [libp2p/go-libp2p-quic-transport#104]
We'll need to:

1. Store session keys in the peerstore.
2. Make sure to _remove_ them when they're used (to prevent replay).
3. Enable session resumption.
4. Somehow give the _server_ a session resumption key (e.g., via a different protocol) so we can resume connections in the other direction.

We've talked about this a lot but we've never actually filed an issue to track it.

**Comment by marten-seemann** [comment]
> 1. Store session keys in the peerstore.

Yes.

> 2. Make sure to remove them when they're used (to prevent replay).

quic-go (or rather, [qtls](https://github.com/marten-seemann/qtls), quic-go's TLS library) does that for us, if and only if 0-RTT is attempted. It's ok to reuse a session ticket if you're not doing 0-RTT.

> 3. Enable session resumption.

... and enable 0-RTT

> 4. Somehow give the server a session resumption key (e.g., via a different protocol) so we can resume connections in the other direction.

Unfortunately, TLS session tickets are only sent from the server to the client.

> We've talked about this a lot but we've never actually filed an issue to track it.

We'll have to be careful what to send in 0-RTT, since 0-RTT data is replayable. It's probably safe to use it for multistream negotiations, but I wouldn't want to send any other application data.

**Comment by Stebalien** [comment]
> Unfortunately, TLS session tickets are only sent from the server to the client.

Yes, but we can generate and send one back in a side-channel (new stream). It shouldn't matter _how_ the server gets the ticket.

> We'll have to be careful what to send in 0-RTT, since 0-RTT data is replayable. It's probably safe to use it for multistream negotiations, but I wouldn't want to send any other application data.

That's why we remove the session key (on the _receiving_ side) once used (2). An attacker could _delay_ the message, I guess. But we can definitely prevent it from being replayed.

**Comment by marten-seemann** [comment]
> > Unfortunately, TLS session tickets are only sent from the server to the client.
> 
> Yes, but we can generate and send one back in a side-channel (new stream). It shouldn't matter _how_ the server gets the ticket.

I'm more worried about the security properties of this. The TLS security analysis doesn't include session tickets sent from client to server (and the specification doesn't even define how you'd generate one).

> > We'll have to be careful what to send in 0-RTT, since 0-RTT data is replayable. It's probably safe to use it for multistream negotiations, but I wouldn't want to send any other application data.
> 
> That's why we remove the session key (on the _receiving_ side) once used (2). An attacker could _delay_ the message, I guess. But we can definitely prevent it from being replayed.

Most implementations, including the Go implementation, encode all the state into the session ticket itself, and don't keep any state on the server side. Clearly, this isn't a protocol requirement, but it would be another major change we'll have to make to the Go TLS library.

**Comment by Stebalien** [comment]
> I'm more worried about the security properties of this. The TLS security analysis doesn't include session tickets sent from client to server (and the specification doesn't even define how you'd generate one).

It's just an encrypted cookie. We'll have think about how we want to do this, but there's no theoretical reason we _can't_.

(note: we can also tackle this later)

> Most implementations, including the Go implementation, encode all the state into the session ticket itself, and don't keep any state on the server side. Clearly, this isn't a protocol requirement, but it would be another major change we'll have to make to the Go TLS library.

Hm. _That_ is annoying.

**Comment by marten-seemann** [comment]
After discussing this with @whyrusleeping, I wrote a short design doc how we can make 0-RTT replay-safe: https://docs.google.com/document/d/1xBin9zQoutncY4rmCivYcQRMwCsxJ1r3dnFrzdaFkFE/edit?usp=sharing
